bagua/torch_api/contrib/cache_loader.py

```diff
@@ -0,0 +1,138 @@
+    return pickle.dumps(input)
+
+
+def deserialize(input):
+    return pickle.loads(input)
+
+
+class CacheLoader:
+    def __init__(
+        self,
+        backend: str = "redis",
+        dataset_name: str = "",
+        writer_buffer_size: int = 1,
+        **kwargs,
+    ):
+        """
+        `CacheLoader` caches values calculated by an expensive function by theirs keys via :func:`CacheLoader.get` method,
+        so that the values can be retrieved faster next time.
+
+        Internally, values are indexed by ``"{dataset_name}_{key}"`` and saved in a distributed Key-Value
+        store, where ``dataset_name`` is specified on initializing, and ``key`` is the argument in :func:`CacheLoader.get`.
+
+        By default, `CacheLoader` uses :class:`RedisStore` as its backend distributed Key-Value store implementation. It
+        supports using a list of existing redis servers or spawning new redis servers. See also
+        :class:`bagua.torch_api.contrib.utils.redis_store.RedisStore`. Parameters for `RedisStore` can be provided here in
+        ``**kwargs``.
+
+        Args:
+            backend(str): Backend distributed Key-Value store implementation. Can be ``"redis"``.
+            dataset_name(str): Name of the dataset. Default ``""``.
+            writer_buffer_size(int): Number of samples to collect before writing to the backend Key-Value store.
+                Useful for improving the backend throughput.
+
+        Example::
+            To use a list of existing redis servers for the "redis" backend:
+
+            >>> from bagua.torch_api.contrib import CacheLoader
+            >>>
+            >>> hosts = [{"host": "192.168.1.0", "port": "7000"}, {"host": "192.168.1.1", "port": "7000"}]
+            >>> loader = CacheLoader(backend="redis", hosts=hosts, cluster_mode=True, dataset_name="test")
+            >>>
+            >>> loader.get(index, lambda x: items[x])
+
+            To spawn new redis servers for the "redis" backend:
+
+            >>> loader = CacheLoader(backend="redis", hosts=None, cluster_mode=True, capacity_per_node=100000000)
+
+        .. note::
+            ``CacheLoader``s with the same ``dataset_name`` will reuse and overwrite each other's cache.
+            Use different ``dataset_name`` if this is not desired.
+
+        """
+
+        self.backend = backend
+        self.dataset_name = dataset_name
+
+        if backend == "redis":
+            from .utils.redis_store import RedisStore
+
+            self.store = RedisStore(**kwargs)
+        else:
+            raise ValueError('Invalid backend, only support "redis" currently')
+
+        self.fetcher = BatchFetcher(self.store, 1, writer_buffer_size)
+
+    def get(self, key, load_fn):
+        """
+        Returns the value associated with key in cache, use ``load_fn`` to create the entry if the key does not exist
+        in the cache. ``load_fn`` is a function taking ``key`` as its argument, and returning corresponding value to
```

Comment:
~~~
   Verbatim in rst (readthedocs doc format) should be ``something`` instead of `something`
~~~
Fix other places